flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java

```diff
@@ -116,20 +116,21 @@ protected void finishTask() throws Exception {
     public CompletableFuture<Boolean> triggerCheckpointAsync(
             CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions) {
         if (!isExternallyInducedSource) {
-            return super.triggerCheckpointAsync(checkpointMetaData, checkpointOptions);
+            if (checkpointOptions.getCheckpointType().shouldDrain()) {
+                return triggerStopWithSavepointWithDrain(checkpointMetaData, checkpointOptions);
+            } else {
+                return super.triggerCheckpointAsync(checkpointMetaData, checkpointOptions);
+            }
         } else {
             return CompletableFuture.completedFuture(isRunning());
         }
     }
 
-    @Override
-    protected CompletableFuture<Void> stopIfSourceMailboxAction() throws Exception {
+    private CompletableFuture<Boolean> triggerStopWithSavepointWithDrain(
```

Comment:
`triggerStopWithSavepointWithDrainAsync`

Reply:
Really sorry, about that. I updated that a few minutes before your review. I forgot to do that after I reverted the part where I moved it to the `performCheckpoint`. Really sorry.